crates/matrix-sdk/src/encryption/recovery/mod.rs

```diff
@@ -0,0 +1,526 @@
+        }
+
+        self.client.add_event_handler(Self::default_key_event_handler);
+        self.client.add_event_handler(Self::secret_send_event_handler);
+
+        Ok(())
+    }
+
+    async fn are_backups_marked_as_disabled(&self) -> Result<bool> {
+        Ok(self
+            .client
+            .account()
+            .fetch_account_data(BackupDisabledContent::event_type())
+            .await?
+            .map(|event| {
+                event
+                    .deserialize_as::<BackupDisabledContent>()
+                    .map(|event| event.disabled)
+                    .unwrap_or(false)
+            })
+            .unwrap_or(false))
+    }
+
+    async fn mark_backup_as_enabled(&self) -> Result<()> {
+        self.client.account().set_account_data(BackupDisabledContent { disabled: false }).await?;
+
+        Ok(())
+    }
+
+    async fn check_recovery_state(&self) -> Result<RecoveryState> {
+        Ok(if self.client.encryption().secret_storage().is_enabled().await? {
+            if self.all_known_secrets_available().await? {
+                RecoveryState::Enabled
+            } else {
+                RecoveryState::Incomplete
+            }
+        } else {
+            RecoveryState::Disabled
+        })
+    }
+
+    async fn update_recovery_state(&self) -> Result<()> {
+        let new_state = self.check_recovery_state().await?;
+        self.client.inner.recovery_state.set(new_state);
+
+        Ok(())
+    }
+
+    async fn update_recovery_state_no_fail(&self) {
+        if let Err(e) = self.update_recovery_state().await {
+            error!("Coulnd't update the recovery state: {e:?}");
+        }
+    }
+
+    #[instrument]
+    async fn secret_send_event_handler(_: ToDeviceSecretSendEvent, client: Client) {
+        client.encryption().recovery().update_recovery_state_no_fail().await;
+    }
+
+    #[instrument]
+    async fn default_key_event_handler(_: SecretStorageDefaultKeyEvent, client: Client) {
+        client.encryption().recovery().update_recovery_state_no_fail().await;
+    }
+
+    #[instrument]
+    pub(crate) async fn update_state_after_backup_disabling(&self) {
+        // TODO: This is quite ugly, this method is called by the backups subsystem.
+        // Backups shouldn't depend on recovery, recovery should listen to the
+        // backup state change.
+        self.update_recovery_state_no_fail().await;
+    }
+
+    #[instrument]
+    pub(crate) async fn update_state_after_keys_query(&self, response: &get_keys::v3::Response) {
+        if let Some(user_id) = self.client.user_id() {
+            if response.master_keys.contains_key(user_id) {
+                // TODO: This is unnecessarily expensive, we could let the crypto crate notify
+                // us that our private keys got erased... Butt, the OlmMachine
+                // gets recreated and... You know the drill by now...
+                self.update_recovery_state_no_fail().await;
+            }
+        }
+    }
```

Comment:
Are these TODOs meant to be addressed before this is merged?

Reply:
No, sadly these items require either the multiprocess refactoring that @bnjbvr is doing to be finished, or some more hacks in the main crate so we deal with the fact that the multiprocess support drops recreates the `OlmMachine`.

Reply:
Would it make sense to add a note in [this issue](https://github.com/matrix-org/matrix-rust-sdk/issues/2624) (under "the final touch"), or reference the issue number in the code (so we can grep the issue number)? That way, we could get back to this later, in this wonderful future where the OlmMachine doesn't get recreated.

Reply:
Great idea, thanks for reminding me.